File: 3DEngine/Delaunay/Tetrahedron.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Delaunay {
    public class Tetrahedron {
        // Store and order the four vertices
        public Point3d[] vertices;
        public Point3d o;      // External center of ¥
        public double r;      // External radius of ¥

        public Tetrahedron(Point3d[] v) {
            this.vertices = v;
            getCenterCircumcircle();
        }

        public Tetrahedron(Point3d v1, Point3d v2, Point3d v3, Point3d v4) {
            this.vertices = new Point3d[4];
            vertices[0] = v1;
            vertices[1] = v2;
            vertices[2] = v3;
            vertices[3] = v4;
            getCenterCircumcircle();
        }

        public bool equals(Tetrahedron t) {
            int count = 0;
            foreach(Point3d p1 in this.vertices) {
                foreach(Point3d p2 in t.vertices) {
                    if(p1.X == p2.X && p1.Y == p2.Y && p1.Z == p2.Z) {
                        count++;
                    }
                }
            }
            if(count == 4) return true;
            return false;
        }

        public Line[] getLines() {
            Point3d v1 = vertices[0];
            Point3d v2 = vertices[1];
            Point3d v3 = vertices[2];
            Point3d v4 = vertices[3];

            Line[] lines = new Line[6];

            lines[0] = new Line(v1, v2);
            lines[1] = new Line(v1, v3);
            lines[2] = new Line(v1, v4);
            lines[3] = new Line(v2, v3);
            lines[4] = new Line(v2, v4);
            lines[5] = new Line(v3, v4);
            return lines;
        }

        // 外接円も求めちゃう
        private void getCenterCircumcircle() {
            Point3d v1 = vertices[0];
            Point3d v2 = vertices[1];
            Point3d v3 = vertices[2];
            Point3d v4 = vertices[3];

            double[][] A = new double[][] {
                                new double[]{v2.X - v1.X, v2.Y-v1.Y, v2.Z-v1.Z},
                                new double[]{v3.X - v1.X, v3.Y-v1.Y, v3.Z-v1.Z},
                                new double[]{v4.X - v1.X, v4.Y-v1.Y, v4.Z-v1.Z}
                            };
            double[] b = {
                                0.5 * (v2.X*v2.X - v1.X*v1.X + v2.Y*v2.Y - v1.Y*v1.Y + v2.Z*v2.Z - v1.Z*v1.Z),
                                0.5 * (v3.X*v3.X - v1.X*v1.X + v3.Y*v3.Y - v1.Y*v1.Y + v3.Z*v3.Z - v1.Z*v1.Z),
                                0.5 * (v4.X*v4.X - v1.X*v1.X + v4.Y*v4.Y - v1.Y*v1.Y + v4.Z*v4.Z - v1.Z*v1.Z)
                            };

            double[] x = new double[3];
            if(gauss(A, b, x) == 0) {
                o = null;
                r = -1;
            } else {
                o = new Point3d((double)x[0], (double)x[1], (double)x[2]);
                r = o.Distance(v1);
            }
        }

        // Solution of equations by LU decomposition
        private double lu(double[][] a, int[] ip) {
            int n = a.Length;
            double[] weight = new double[n];

            for(int k = 0; k < n; k++) {
                ip[k] = k;
                double u = 0;
                for(int j = 0; j < n; j++) {
                    double t = Math.Abs(a[k][j]);
                    if(t > u) u = t;
                }
                if(u == 0) return 0;
                weight[k] = 1 / u;
            }

            double det = 1;
            for(int k = 0; k < n; k++) {
                double u = -1;
                int m = 0;
                for(int i = k; i < n; i++) {
                    int ii = ip[i];
                    double t = Math.Abs(a[ii][k]) * weight[ii];
                    if(t > u) { u = t; m = i; }
                }
                int ik = ip[m];
                if(m != k) {
                    ip[m] = ip[k]; ip[k] = ik;
                    det = -det;
                }
                u = a[ik][k]; det *= u;
                if(u == 0) return 0;
                for(int i = k + 1; i < n; i++) {
                    int ii = ip[i];
                    double t = (a[ii][k] /= u);
                    for(int j = k + 1; j < n; j++) a[ii][j] -= t * a[ik][j];
                }
            }
            return det;
        }

        private void solve(double[][] a, double[] b, int[] ip, double[] x) {
            int n = a.Length;
            for(int i = 0; i < n; i++) {
                int ii = ip[i];
                double t = b[ii];
                for(int j = 0; j < i; j++) t -= a[ii][j] * x[j];
                x[i] = t;
            }
            for(int i = n - 1; i >= 0; i--) {
                double t = x[i];
                int ii = ip[i];
                for(int j = i + 1; j < n; j++) t -= a[ii][j] * x[j];
                x[i] = t / a[ii][i];
            }
        }

        private double gauss(double[][] a, double[] b, double[] x) {
            int[] ip = new int[a.Length];
            double det = lu(a, ip);

            if(det != 0) { solve(a, b, ip, x); }
            return det;
        }
    }
}

File: 3DEngine/Classes/Delaunay/Tetrahedron.vb
Namespace Delaunay
    Public Class Tetrahedron
        Implements IEquatable(Of Tetrahedron)

        Private mVertices(4 - 1) As Point3d
        Private mCenter As Point3d
        Private mRadius As Double

        Public Sub New(p() As Point3d)
            mVertices = p
            GetCenterCircumcircle()
        End Sub

        Public Sub New(p1 As Point3d, p2 As Point3d, p3 As Point3d, p4 As Point3d)
            mVertices = {p1, p2, p3, p4}
            GetCenterCircumcircle()
        End Sub

        Public ReadOnly Property Vertices As Point3d()
            Get
                Return mVertices
            End Get
        End Property

        Public ReadOnly Property Center As Point3d
            Get
                Return mCenter
            End Get
        End Property

        Public ReadOnly Property Radius As Double
            Get
                Return mRadius
            End Get
        End Property

        Public ReadOnly Property Lines() As Line()
            Get
                Return New Line() {New Line(mVertices(0), mVertices(1)),
                                   New Line(mVertices(0), mVertices(2)),
                                   New Line(mVertices(0), mVertices(3)),
                                   New Line(mVertices(1), mVertices(2)),
                                   New Line(mVertices(1), mVertices(3)),
                                   New Line(mVertices(2), mVertices(3))}
            End Get
        End Property

        Private Sub GetCenterCircumcircle()
            Dim v1 As Point3d = mVertices(0)
            Dim v2 As Point3d = mVertices(1)
            Dim v3 As Point3d = mVertices(2)
            Dim v4 As Point3d = mVertices(3)

            Dim a()() As Double = New Double()() {
                    New Double() {v2.X - v1.X, v2.Y - v1.Y, v2.Z - v1.Z},
                    New Double() {v3.X - v1.X, v3.Y - v1.Y, v3.Z - v1.Z},
                    New Double() {v4.X - v1.X, v4.Y - v1.Y, v4.Z - v1.Z}
                }

            Dim b() As Double = {
                    0.5 * (v2.X * v2.X - v1.X * v1.X + v2.Y * v2.Y - v1.Y * v1.Y + v2.Z * v2.Z - v1.Z * v1.Z),
                    0.5 * (v3.X * v3.X - v1.X * v1.X + v3.Y * v3.Y - v1.Y * v1.Y + v3.Z * v3.Z - v1.Z * v1.Z),
                    0.5 * (v4.X * v4.X - v1.X * v1.X + v4.Y * v4.Y - v1.Y * v1.Y + v4.Z * v4.Z - v1.Z * v1.Z)
                }

            Dim x(3 - 1) As Double
            If Gauss(a, b, x) = 0 Then
                mCenter = Nothing
                mRadius = -1
            Else
                mCenter = New Point3d(x(0), x(1), x(2))
                mRadius = mCenter.Distance(v1)
            End If
        End Sub

        Private Function Gauss(a()() As Double, b() As Double, x() As Double) As Double
            Dim ip(a.Length - 1) As Integer
            Dim det As Double = LU(a, ip)

            If det <> 0.0 Then Solve(a, b, ip, x)
            Return det
        End Function

        ' Solution of equations by LU decomposition
        ' https://en.wikipedia.org/wiki/LU_decomposition
        Private Function LU(a()() As Double, ip() As Integer) As Double
            Dim n As Integer = a.Length - 1
            Dim weight(n) As Double
            Dim u As Double
            Dim t As Double

            For k As Integer = 0 To n
                ip(k) = k
                u = 0.0
                For j As Integer = 0 To n
                    t = Math.Abs(a(k)(j))
                    If t > u Then u = t
                Next
                If u = 0.0 Then Return 0.0
                weight(k) = 1 / u
            Next

            Dim m As Integer
            Dim ii As Integer
            Dim ik As Integer
            Dim det As Double = 1.0
            For k As Integer = 0 To n
                u = -1.0
                m = 0
                For i As Integer = k To n
                    ii = ip(i)
                    t = Math.Abs(a(ii)(k)) * weight(ii)
                    If t > u Then
                        u = t
                        m = i
                    End If
                Next

                ik = ip(m)
                If m <> k Then
                    ip(m) = ip(k)
                    ip(k) = ik
                    det = -det
                End If

                u = a(ik)(k)
                det *= u
                If u = 0.0 Then Return 0.0
                For i As Integer = k + 1 To n
                    ii = ip(i)
                    a(ii)(k) /= u
                    t = a(ii)(k)
                    For j As Integer = k + 1 To n
                        a(ii)(j) -= t * a(ik)(j)
                    Next
                Next
            Next

            Return det
        End Function

        Private Sub Solve(a()() As Double, b() As Double, ip() As Integer, x() As Double)
            Dim n As Integer = a.Length - 1
            Dim ii As Integer
            Dim t As Double

            For i As Integer = 0 To n
                ii = ip(i)
                t = b(ii)
                For j = 0 To i - 1
                    t -= a(ii)(j) * x(j)
                Next
                x(i) = t
            Next

            For i As Integer = n To 0 Step -1
                t = x(i)
                ii = ip(i)
                For j = i + 1 To n
                    t -= a(ii)(j) * x(j)
                Next
                x(i) = t / a(ii)(i)
            Next
        End Sub

        Public Shared Operator =(t1 As Tetrahedron, t2 As Tetrahedron) As Boolean
            Dim counter As Integer = 0
            For Each p1 In t1.Vertices
                For Each p2 In t2.Vertices
                    If p1.X = p2.X AndAlso p1.Y = p2.Y AndAlso p1.Z = p2.Z Then counter += 1
                Next
            Next

            Return counter = 4
        End Operator

        Public Shared Operator <>(t1 As Tetrahedron, t2 As Tetrahedron) As Boolean
            Return Not (t1 = t2)
        End Operator

        Public Shadows Function Equals(other As Tetrahedron) As Boolean Implements IEquatable(Of Tetrahedron).Equals
            Return Me = other
        End Function
    End Class
End Namespace
